docs/operations_guide/Operations.md

```diff
@@ -0,0 +1,623 @@
+* Error rates - are there any problems with execution
+
+
+##### Checking source lag
+
+There are two types of metrics that can give useful information about potential Kafka lags
+
+
+
+* How large is lag on Kafka source partitions
+* How much time has passed since the last processed event
+
+Please note that:
+
+
+
+* Kafka lag is reported with some delay (offsets are not committed immediately). It’s normal to have temporary spikes.
+* Large differences between slots in lag may point to data skew or e.g. problematic data for some keys
+
+
+#### Technical metrics
+
+There are also some more technical metrics that should be looked at. While they are not directly connected to scenario performance or scenario errors, they can point to some problems with either scenario configuration, its performance or problems with the cluster itself.
+
+They are presented in the “Scenario health” row in the scenario dashboard.
+
+
+![Scenario health](img/scenario_health.png "scenario health")
+
+
+Things to watch out for (they are usually a sign of a problem):
+
+
+
+* Low uptime metric with a high number of restarts
+* Long checkpoint time (e.g. many minutes for checkpoint size lower than many GB)
+* A non-zero number of failed checkpoints
+
+
+### Handling typical scenario errors
+
+When you see errors in monitoring, like on screen below:
+
+
+
+![Scenario errors](img/scenario_errors.png "typical errors")
+
+
+it means that there were failures during scenario execution - most often they are caused by errors during SpeL expression evaluation, which couldn’t be detected during validation.
+
+The most common problems are:
+
+
+
+* NullPointerException - please see SpEL documentation [link Safe Navigation] to see how to check if some value is null
+* Conversion errors - e.g. #NUMBER.toNumber - this usually indicates problem with the source data
+
+To see the exact error you can look at (depending on the configuration of the Exception handler):
+
+
+
+* Logs of TaskManager (be sure to check all of them - e.g. using ELK):
+
+
+```
+2021-06-24 07:43:32,071 INFO  pl.touk.nussknacker.engine.spel.SpelExpression               [] - Expression evaluation failed. Original #client.status != "GOLD", ctxId: DetectLargeTransactions-kafka-registry-typed-json-1-3847, message: EL1021E: A problem occurred whilst attempting to access the property 'status': 'Cannot invoke method/property status on null object'
+
+```
+
+
+
+* Contents of configured errors topic on Kafka:
+
+
+```
+{
+  "processName": "DetectLargeTransactions",
+  "nodeId": "only large ones",
+  "message": "Expression [20 / (#input.amount % 4) > 0] evaluation failed, message: / by zero",
+  "exceptionInput": "20 / (#input.amount % 4) > 0",
+  "inputEvent": null,
+  "stackTrace": "pl.touk.nussknacker.engine.spel.SpelExpressionEvaluationException:... 34 more",
+  "timestamp": 1624863781881,
+  "host": "0bda71c51449",
+  "additionalData": {}
+}
+```
+
+
+
+
+
+### Tuning scenario performance
```

Comment:
We need an additional subsection on "managing Flink partitioning".  How does NK decide to assign key (keyBy argument) in the operators? Can we (using NK) affect how NK instructs Flink to choose the key? Does the key get changed during the scenario execution? 

It is only now that I understood why some fields in the Component configuration forms are called _keyBy_.  As a side effect of the aggregation window we (also) change(?) the key ... This cannot be a secret - although may be an advanced topic. BTW, we have agreed to change the keyBy label to groupBy label in the aggregate forms; I just detected this field in the Delay component. Is it somewhere else?